### PR TITLE
chore: glue HttpService with MetricService continue

### PR DESCRIPTION
**Pull request by Kazim1101**
[Extension] Glue HttpService with MetricService [#3](https://github.com/YunaBraska/nano/issues/3) 

## Motivation
glue HttpService with MetricService

### CODE REVIEW

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 1]
```diff
@@ -66,6 +71,56 @@ public void onEvent(final Event event) {
...
             .ifPresent(EVENT_APP_LOG_QUEUE, LogQueue.class, logger::logQueue);
+        addMetricsEndpoint(event);
+
+    }
+
+    protected void addMetricsEndpoint(Event event) {
+        event
+            .ifPresent(EVENT_HTTP_REQUEST, HttpObject.class, request ->
+            {
+                request.pathMatch(Config.getMetricsBaseUrl() + Config.getPrometheusBaseUrl());
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:82]: It should be `Config.getMetricsBaseUrl() + "prometheus"` or `Config.getPrometheusUrl()` if it's set.
Also please take care if the first and last `/` in `Config.getMetricsBaseUrl()` and `Config.getPrometheusUrl()`. Users could add or forget them:
* NO: metrics/prometheus
* NO: metrics/prometheus/
* NO: /metrics/prometheus/
* YES: /metrics/prometheus

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 2]
```diff
@@ -78,49 +80,49 @@ public void onEvent(final Event event) {
...
-                    final String response = metrics.prometheus();
-                    request.response()
-                        .statusCode(200)
-                        .body(metrics.prometheus())
-                        .headerMap(Map.of(HttpHeaders.CONTENT_TYPE, ContentType.TEXT_PLAIN)).send(event);
-
-            })
+                Optional.ofNullable(prometheusPath)
+                    .filter( request::pathMatch)
+                    .filter(path -> request.isMethodGet())
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:85]: can be replaced with lambda

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 3]
```diff
@@ -36,13 +40,11 @@ public MetricService() {
...
     public void start(final Supplier<Context> contextSupplier) {
         isReady.set(false, true, run -> updateSystemMetrics());
-        //addlistener for event
-
-        // metrics.prometheus(); this will the response body
-        // GET call
-        // http://localhost:8080/metrics thinks of good word
-        // status code 200
-
+        Optional<String> basePath = contextSupplier.get().getOpt(String.class, Config.CONFIG_METRIC_SERVICE_BASE_PATH.id());
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:43]: move this inside the `isReady.set` beckets as the ready state is set to true before the paths are evaluated

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 4]
```diff
@@ -28,6 +28,10 @@
 @SuppressWarnings({"unused", "UnusedReturnValue"})
 public class MetricService extends Service {
     private final MetricCache metrics = new MetricCache();
+    private String prometheusPath;
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:31]: no `private`, use protected. We are not hiding something from the developers. When someone wants to override something, they are free to do it

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 5]
```diff
@@ -28,6 +28,10 @@
...
+    private String dynamoPath;
+    private String influx;
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:33]: provide getters

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 6]
```diff
@@ -78,49 +80,49 @@ public void onEvent(final Event event) {
...
-                request.isMethodGet();
-                final String response = metrics.influx();
-                request.response()
-                    .statusCode(200)
-                    .body(metrics.influx())
-                    .headerMap(Map.of(HttpHeaders.CONTENT_TYPE, ContentType.TEXT_PLAIN)).send(event);
-
-            })
+                Optional.ofNullable(dynamoPath)
+                    .filter( request::pathMatch)
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:95]: Some spacing issues, reformat the class.

File: src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java [hunk 7]
```diff
@@ -12,27 +12,19 @@
...
-        assertThat(((HttpObject) event.payload()).path()).isNotEqualTo("/metrics/prometheus");
-        assertThat(event.response(HttpObject.class).statusCode()).isEqualTo(200);
-    }
-
-    @Test
-    void testAnyMetricsEndpoint() {
-
-        Event event= new Event(EVENT_HTTP_REQUEST, Context.createRootContext(), new HttpObject().methodType(HttpMethod.GET).path("/metrics/influx"), null);
+    void testMetricsEndpoint() {
+        String influxPath = "/metrics/influx";
```
Comment (YunaBraska) [inline src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java:16]: We need Integration Tests here as we want to prove the connection between `HttpService` and `MetricService`.
UnitTests should be only used for testing dummy model structure.

### Test with different configurations:
* without basePath
* with basePath
* with prometheus path
* with base path and prometheus path

Integration Test could look like:
```
final Nano nano = new Nano(Map.of(CONFIG_LOG_LEVEL, TEST_LOG_LEVEL), new MetricService(), new HttpService());

final HttpObject result = new HttpObject()
  .methodType(HttpMethod.GET)
  .path("/metrics/prometheus")
  .send(nano.newContext(MetricServiceTest.class));

// TODO: assertions
assertThat(result).isNotNull();
assertThat(result.bodyAsString()).containsString("AA", "BB",...)

nano.shutdown(this.getClass());
assertThat(service.stopCount()).isEqualTo(1);
```

File: src/test/resources/application.properties [hunk 8]
```diff
@@ -6,3 +6,4 @@ test.placeholder.fallback.empty=${invalid_key:}
 test.placeholder.value=${placeholder_value:fallback should not be used}
 test.placeholder.key_empty=${:fallback should be used 2}
 placeholder_value=used placeholder value
+metrics_base_url=/metrics
```
Comment (YunaBraska) [inline src/test/resources/application.properties:9]: That's not needed for the tests

File: src/main/java/berlin/yuna/nano/core/model/Config.java [hunk 9]
```diff
@@ -30,8 +30,14 @@ public enum Config {
     CONFIG_HTTP_CLIENT_MAX_RETRIES("app_service_http_max_retries", "Maximum number of retries for the HTTP client (see " + HttpClient.class.getSimpleName() + ")"),
     CONFIG_HTTP_CLIENT_CON_TIMEOUT_MS("app_service_http_con_timeoutMs", "Connection timeout in milliseconds for the HTTP client (see " + HttpClient.class.getSimpleName() + ")"),
     CONFIG_HTTP_CLIENT_READ_TIMEOUT_MS("app_service_http_read_timeoutMs", "Read timeout in milliseconds for the HTTP client (see " + HttpClient.class.getSimpleName() + ")"),
-    CONFIG_HTTP_CLIENT_FOLLOW_REDIRECTS("app_service_http_follow_redirects", "Follow redirects for the HTTP client (see " + HttpClient.class.getSimpleName() + ")");
-    ;
+    CONFIG_HTTP_CLIENT_FOLLOW_REDIRECTS("app_service_http_follow_redirects", "Follow redirects for the HTTP client (see " + HttpClient.class.getSimpleName() + ")"),
+
+    // ########## METRIC SERVICES ##########
+    CONFIG_METRIC_SERVICE_BASE_PATH("metrics_base_url", "Base path for the metric service"),
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/core/model/Config.java:36]: Use prefix `app_service_`

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java (hunk 2 above)
Comment (Kazim1101) [reply src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:90]: `filter(HttpObject::isMethodGet)`
Its complaining about "Non-static method cannot be referenced from a static context", else have to change `isMethodGet` static

File: src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java [hunk 10]
```diff
@@ -281,4 +343,20 @@ private long estimateMetricSize(final MetricCache.Metric<?> metric, final long n
     private long estimateStringSize(final String string) {
         return 24 + (long) string.length() * 2; // String object overhead + 2 bytes per character
     }
+
+    public String getPrometheusPath() {
```
Comment (YunaBraska) [inline src/main/java/berlin/yuna/nano/services/metric/logic/MetricService.java:347]: remember: no `get` and no `set` prefix

File: src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+import static berlin.yuna.nano.core.model.Config.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MetricServiceTest {
+
+    protected static String serverUrl = "http://localhost:";
+
+    @Test
+    void metricEndpointsWithoutBasePath() {
+        Nano nano = new Nano(Map.of(CONFIG_LOG_LEVEL, LogLevel.INFO, CONFIG_LOG_FORMATTER, "console"), new LogQueue(), new MetricService(), new HttpService());
```
Comment (YunaBraska) [inline src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java:22]: Remove unnecessary setup (`YAGNI`)
* `CONFIG_LOG_FORMATTER, "console"` is not required for this test
* `new LogQueue()` is not required for this test

File: src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+        final HttpObject result = new HttpObject()
+            .methodType(HttpMethod.GET)
+            .path(serverUrl + nano.service(HttpService.class).port() + "/metrics/prometheus")
```
Comment (YunaBraska) [inline src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java:26]: Nice catch with the port and services 👍

File: src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java (hunk 11 above)
Comment (YunaBraska) [inline src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java:22]: Use global test LogLevel: `Nano(Map.of(CONFIG_LOG_LEVEL, TEST_LOG_LEVEL)`. Makes it easier to switch log levels while debugging

File: src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            .send(nano.newContext(MetricServiceTest.class));
+
+        assertThat(result).isNotNull();
+        assertThat(result.statusCode()).isEqualTo(200);
+        assertThat(nano.stop(MetricServiceTest.class).waitForStop().isReady()).isFalse();
+    }
+
+    @Test
+    void metricEndpointsWithBasePath() {
+        Nano nano = new Nano(Map.of(CONFIG_LOG_LEVEL, LogLevel.INFO, CONFIG_LOG_FORMATTER, "console", CONFIG_METRIC_SERVICE_BASE_PATH, "/metrics"), new LogQueue(), new MetricService(), new HttpService());
```
Comment (YunaBraska) [inline src/test/java/berlin/yuna/nano/services/metric/logic/MetricServiceTest.java:66]: Use a different path than the default path ;)

**Review by YunaBraska — APPROVED** [review]

**Pull request merged**